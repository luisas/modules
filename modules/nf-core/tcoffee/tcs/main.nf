process TCOFFEE_TCS {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/t-coffee_pigz:91ac7e26b23bb246':
        'community.wave.seqera.io/library/t-coffee_pigz:7d1373a24f76afe6' }"

    input:
    tuple val(meta) , path(msa)
    tuple val(meta2), path(lib)

    output:
    tuple val(meta), path("*.tcs")   , emit: tcs
    tuple val(meta), path("*.scores"), emit: scores
    path "versions.yml"              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args          = task.ext.args ?: ''
    def prefix        = task.ext.prefix ?: "${meta.id}"
    def lib_arg       = lib ? "-lib ${lib}" : ""
    def header        = meta.keySet().join(",")
    def values        = meta.values().join(",")
    def unzipped_name = msa.toString() - '.gz'
    """
    export TEMP='./'
    filename=${msa}
    if [[ \$(basename $msa) == *.gz ]] ; then
        unpigz -f $msa
        filename=${unzipped_name}
    fi

    # Bad hack to circumvent t_coffee bug
    # Issue described already in: https://github.com/cbcrg/tcoffee/issues/3
    # Add an A in front of filename if the file begins with A
    first_letter_filename=\${filename:0:1}
    if [ "\$first_letter_filename" == "A" ]; then input="A"\$filename; else input=\$filename;  fi

    t_coffee -infile \$input \
        -evaluate -output=score_ascii \
        ${lib_arg} \
        -outfile ${prefix}.tcs

    # Add metadata info to output file
    echo "${header},TCS" > "${prefix}.scores"

    # Add values
    scores=\$(grep 'SCORE=' ${prefix}.tcs | cut -d '=' -f 2 )
    echo "${values},\$scores" >> "${prefix}.scores"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tcoffee: \$( t_coffee -version | awk '{gsub("Version_", ""); print \$3}')
        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\\w*//' ))
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    # Otherwise, tcoffee will crash when calling its version
    export TEMP='./'
    touch ${prefix}.tcs
    touch ${prefix}.scores

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        tcoffee: \$( t_coffee -version | awk '{gsub("Version_", ""); print \$3}')
        pigz: \$(echo \$(pigz --version 2>&1) | sed 's/^.*pigz\\w*//' ))
    END_VERSIONS
    """
}
