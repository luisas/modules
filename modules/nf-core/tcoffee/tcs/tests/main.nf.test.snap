{
    "tcs - seatoxin - compressed": {
        "content": [
            [
                "versions.yml:md5,696cea7c0ca72bf5aae47d05df2276d8"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:02:50.049720694"
    },
    "tcs - seatoxin - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.tcs:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.scores:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,696cea7c0ca72bf5aae47d05df2276d8"
                ],
                "scores": [
                    [
                        {
                            "id": "test"
                        },
                        "test.scores:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "tcs": [
                    [
                        {
                            "id": "test"
                        },
                        "test.tcs:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,696cea7c0ca72bf5aae47d05df2276d8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:03:13.55612748"
    },
    "tcs - seatoxin - lib": {
        "content": [
            [
                "versions.yml:md5,696cea7c0ca72bf5aae47d05df2276d8"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:03:03.442861443"
    },
    "tcs - seatoxin": {
        "content": [
            [
                "versions.yml:md5,696cea7c0ca72bf5aae47d05df2276d8"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-19T08:02:34.071353672"
    }
}